### DISCUSSION TITLE
Couldn't connect my postgress while installing medusa.

### DISCUSSION THREAD

**Discussion by Fardeen-Awais** [medusajs/medusa#4743] [Q&A] [answered]
I had already downloaded my posgresql in my computer and created the database but when i installing medusa it is not working. I'm 100% sure i'm providing the right password

`PS G:\Practices\04_Medusa_experiment> yarn create medusa-app
yarn create v1.22.19
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Installed "create-medusa-app@1.0.4" with binaries:
      - create-medusa-app
? What's the name of your project? my-medusa-store
? Enter your Postgres username postgres
? Enter your Postgres password [hidden]
Couldn't connect to PostgreSQL. Make sure you have PostgreSQL installed and the credentials you provided are correct.

You can learn how to install PostgreSQL here: https://docs.medusajs.com/development/backend/prepare-environment?os=windows#postgresql

If you keep running into this issue despite having PostgreSQL installed, please check out our troubleshooting guidelines: https://docs.medusajs.com/troubleshooting/database-error
error Command failed.
Exit code: 1
Command: C:\Users\farde\AppData\Local\Yarn\bin\create-medusa-app
Arguments:
Directory: G:\Practices\04_Medusa_experiment
Output:
`
![image](https://github.com/medusajs/medusa/assets/101203786/8bd8f5f3-ccf0-4f1a-810b-32fef655b4c2)

**Comment by AbhinavChavareTechcopain** [comment]
please check the port of PostgreSQL for medusa it should be 5432.
Thanks

**Comment by srindom** [comment]
cc: @shahednasser

**Comment by shahednasser** [comment, marked as answer]
@Fardeen-Awais with the latest update of `create-medusa-app`, you can set your database url by passing it as an option as follows:

```bash
npx create-medusa-app@latest --db-url postgres:://user:password@localhost:5432/giftstore
```

Make sure to pass the correct username, password, and port (by default it's `5432`).

Please note that with the latest version there's an issue with using `yarn create medusa-app`. We've opened a PR to fix it, so please use npx or pnpm for now.

**Reply by Fardeen-Awais** [reply]
Thanks the problem was in my port. I changed my port that cause error.

**Reply by Fab1en** [reply]
Thanks @shahednasser but there is just a typo : it should be `postgres://user:password@localhost:5432/giftstore` instead of `[3A7hxV5yDaMMuzeg](postgres:://user:password@localhost:5432/giftstore)` (only one colon after postgres)

**Comment by zaidrashid** [comment]
Hello, I have the same issue, but my Postgres is running in Docker.

When I tested my connection via `psql` command, I am able to connect to it. 

![image](https://github.com/medusajs/medusa/assets/5754316/7b62f925-0bd0-4ef5-8b65-9f44167186b7)

However, when I passed the same command to start a medusa project, I ran into this issue.

![image](https://github.com/medusajs/medusa/assets/5754316/b9d6c96c-a78b-4253-8ef2-dfb7b7ecab5b)

I am currently just testing it, so my Postgres user is `user` and its password is `password` and my database is `mydatabase` and the port is `5432` (I did not pass the -p in the `psql` command as that is the default).

Anything that I missed? Thanks!

**Reply by shahednasser** [reply]
Thanks for all the details. Can you try wrapping the database URL with double quotes?

**Reply by zaidrashid** [reply]
Do you mean like this?

`npx create-medusa-app@latest --seed --db-url "postgres:://user:password@localhost:5432/medusa_db"`

![image](https://github.com/medusajs/medusa/assets/5754316/cfe908a2-c59c-45e3-8c8d-c2c2ad8f6b30)

Still does not work.

By the way, I am running on a Windows machine in WSL. I am planning to test it on my Mac tomorrow. But do you think the environment plays a part here?

And I also changed the db name, installed a new postgres container. In case you are wondering why the db name changed.

**Reply by shahednasser** [reply]
can you try adding `?sslmode=require` at the end of the URL? Like this:

```bash
npx create-medusa-app@latest --seed --db-url "postgres:://user:password@localhost:5432/medusa_db?sslmode=require"
```

Typically this option is only necessary if you're connecting to a production environment, but I'm not sure if it's necessary for Docker as well. I'll give it a spin tomorrow with Docker and see if I can reproduce the issue and find a solution. Please make sure that the command can connect to the postgres container with just `localhost`.

If you also test it out on Mac, let me know what the outcome is.

**Reply by zaidrashid** [reply]
So, I've just tested it out on Mac and I am still having the same issue (same error, could not connect to PostgresSQL). 

The same setup, postgres docker on my Mac (M1).

I've also tried appending the `sslmode` and puting double quote, issue still persist.

Did you manage to try setting up `medusa` app with a Postgresql?

**Reply by shahednasser** [reply]
@zaidrashid sorry no. I'll try when I get some time this week and let you know

**Reply by shahednasser** [reply]
@zaidrashid just tested it out, and at first I wasn't able to connect to the database. But then after making sure the port was published, it connected successfully.

Can you make sure you published/exposed the `5432` port of your docker container? If you're running it from the terminal, it can be done with the `-p` option. For example:

```bash
docker run -d -p 5432:5432 --name some-postgres -e POSTGRES_PASSWORD=supersecret postgres
```

From the Docker Desktop you can specify the exposed port as well:

<img width="581" alt="Screenshot 2023-11-14 at 10 56 04 AM" src="https://github.com/medusajs/medusa/assets/27354907/bdb956a4-60b6-496b-b782-a208040ec517">

Let me know if this works out.

**Reply by zaidrashid** [reply]
Hi @shahednasser , sorry for the late reply. I was on holiday and just returned last week. 

I've tried that actually and tried it again. Here's a snapshot of my container. 
![image](https://github.com/medusajs/medusa/assets/5754316/87752c2c-37dd-4dab-bf04-fe78005ee511)

![image](https://github.com/medusajs/medusa/assets/5754316/4784f8de-243e-4cf5-b11e-95c089fea9fb)


But still having issues with the `medusa-create-app` though. Is there some inner logs that I could check by myself when creating the medusa-app? 

![image](https://github.com/medusajs/medusa/assets/5754316/47f36649-f70e-4464-a798-4dd40659a3e1)

**Reply by shahednasser** [reply]
@zaidrashid unfortunately there isn't. Maybe try setting up the Medusa backend following [this guide](https://docs.medusajs.com/development/backend/install)?  It gives you more flexibility during the setup process and maybe the error will be clearer.

**Reply by zaidrashid** [reply]
So what I did just now is to clone the whole [repo](https://github.com/medusajs/medusa) and build it so that I can use the `create-medusa-app` package. I was hoping to catch the same error, and output the error message here,
https://github.com/medusajs/medusa/blob/8f25ed8a10fe23e9342dc3d03545546b4ad4d6da/packages/create-medusa-app/src/utils/create-db.ts#L127-L131

but surprisingly it works on both `develop` branch and [`create-medusa-app@1.2.5` tagged branch](https://github.com/medusajs/medusa/tree/create-medusa-app%401.2.5).

For now, I will explore `medusa` since I managed to create the app, but I will try later setting it up from the guide you provided. Thanks for the support so far. I'll let you know if I have further questions.

**Comment by tomavant** [comment]
I was having this same problem with docker, and realized I had my local posgres for another project still running on 5432, `sudo lsof -i :5432`. I switched my docker port to 5431, and create-medusa-app ran successfully

**Reply by samuic** [reply]
in some cases can be the port, this can be executing other process, please review and list process for kill and start new instalation.
windows
bash:
tasklist | findstr ### -> list number of process of your localhost 
then run 
bash:
taskkill /F /PID ### -> number of process 

works for me

**Comment by kensteele** [comment]
## Issue
> [!NOTE]
> This is a reproduction of bug https://github.com/medusajs/medusa/issues/4999

Running `npx create-medusa-app@latest --with-nextjs-starter` fails with `error: role "postgres" does not exist`

### Environment
Versions:
```markdown
OSX:               14.4 (Sonoma)
NodeJS:            v21.7.1
Homebrew:          4.2.15-32
create-medusa-app: 1.2.7
```
### Steps to reproduce
1. `brew install postgres`
2. `npx create-medusa-app@latest --with-nextjs-starter`
3. Accept the defaults for project name, email, postgres username and postgres password

### Results
* Fails with error: role "postgres" does not exist:
```markdown
npx create-medusa-app@latest --with-nextjs-starter
? What's the name of your project? my-medusa-store
? Enter an email for your admin dashboard user admin@medusa-test.com
? Enter your Postgres username postgres
? Enter your Postgres password [hidden]
Couldn't connect to PostgreSQL because of the following error: error: role "postgres" does not exist.

Make sure you have PostgreSQL installed and the credentials you provided are correct.

You can learn how to install PostgreSQL here: https://docs.medusajs.com/development/backend/prepare-environment?os=macos#postgresql

If you keep running into this issue despite having PostgreSQL installed, please check out our troubleshooting guidelines: https://docs.medusajs.com/troubleshooting/database-error
```

### Expected results
* `npx create-medusa-app@latest --with-nextjs-starter` should succeed
# SOLUTION
* Create a `postgres` role with all privileges: 
> [!IMPORTANT]
>```shell
> psql -U $USER -d postgres -c 'CREATE ROLE "postgres"' \
> -c 'ALTER ROLE "postgres" WITH SUPERUSER INHERIT CREATEROLE CREATEDB LOGIN REPLICATION BYPASSRLS'
>```
* To verify the role was created, you should now see `${USER}` and postgres when dumping the roles:
```shell
psql -U $USER -d postgres -c '\du'
                                    List of roles
 Role name |                         Attributes                         | Member of 
-----------+------------------------------------------------------------+-----------
 ${USER}   | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 postgres  | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
```
* Run `create-medusa-app` again
```shell
npx create-medusa-app@latest --with-nextjs-starter
```

# Further Analysis
> [!NOTE]
> Substitute `${USER}` below with your currently logged in OSX username

`brew install postgres` on OSX installs `postgres@14` and does the following:

* creates a role `${USER}` with no password
* grants the `${USER}` role all privileges
* creates a database `postgres`

`npx create-medusa-app@latest --with-nextjs-starter` appears to assume the database name and the role name are the same, which results in either a database or role error when they don't match.
* when specifying `${USER}` for the `Postgres username`:
```
Couldn't connect to PostgreSQL because of the following error: error: database "${USER}" does not exist.
```
* when specifying `postgres` (or default) for the `Postgres username`:
```
Couldn't connect to PostgreSQL because of the following error: error: role "postgres" does not exist.
```
* Upon inspecting the `create-medusa-app` npm module source in the medusa `v1.20.4` tag, we see that the user is prompted to enter their database username and password, but not the database name.
https://github.com/medusajs/medusa/blob/2524a9dfb0d32d144d9ef3f66856c5b7d56943d8/packages/create-medusa-app/src/utils/create-db.ts#L63-L85
### Potential fixes
1. In `utils/create-db.ts`, prompt the user to specify their database name (with a default of `postgres`):
```ts
      {
        type: "input",
        name: "postgresDatabaseName",
        message: "Enter your Postgres database name",
        default: "postgres",
        validate: (input) => {
          return typeof input === "string" && input.length > 0
        },
      },
    ])

    postgresUsername = answers.postgresUsername
    postgresPassword = answers.postgresPassword
    postgresDatabaseName = answers.postgresDatabaseName
```
2. When the initial db connection with username `postgres` and database name `postgres` fails, automatically attempt to connect using username: `process.env.USER` and database name: `postgres`

**Comment by Sands-45** [comment]
For anyone connecting to external postgres dn please make sure to wrap the db url in double quotes and specify the the ssl mode `npx create-medusa-app@latest --db-url "postgres:://user:password.db?ssl-true`

**Comment by ohmygoobness** [comment]
this is the correct `npx create-medusa-app@latest --db-url "postgres://user:password@localhost:5432/medusadb`

**Comment by njNagaraj** [comment]
C:\Users\Nj\Desktop\Medusa.js>npx create-medusa-app@latest --db-url "postgresql://medusajs_db_user:password@dpg-cukpu0l2ng1s73fvs2kg-a.singapore-postgres.render.com/medusajs_db"
? What's the name of your project? medusa-app
? Would you like to create the Next.js storefront? You can also create it later No
? Enter your Postgres username medusajs_db_user
? Enter your Postgres password [hidden]
? Enter your Postgres user's database name medusajs_db
Couldn't connect to PostgreSQL because of the following error: AggregateError.

Make sure you have PostgreSQL installed and the credentials you provided are correct.

If you keep running into this issue despite having PostgreSQL installed, please check out our troubleshooting guidelines: https://docs.medusajs.com/resources/troubleshooting/database-errors

I'm trying to connect my postgres db from render.com to medusa locally but it's not working. someone help

**Reply by Fab1en** [reply]
It seams that providing `db-url` to the `create-medusa-app` does not work at this time. You can use `--skip-db` to install the project without database connexion, and then enter your database credentials in the `.env` file and use [Medusa CLI](https://docs.medusajs.com/resources/medusa-cli/commands/db) to setup the database afterward : `npx medusa db:setup `
